File: backend/src/services/governanceService.js
/**
 * Governance Service - DAO Operations for Panchayat
 * 
 * Handles on-chain governance operations including:
 * - Creating proposals
 * - Voting on proposals
 * - Executing passed proposals
 * - Querying governance state
 */

const { ethers } = require('ethers');
const config = require('../config');
const logger = require('../utils/logger');

// Governor Contract ABI (OpenZeppelin Governor compatible)
const GOVERNOR_ABI = [
  // Proposal functions
  'function propose(address[] targets, uint256[] values, bytes[] calldatas, string description) returns (uint256)',
  'function queue(address[] targets, uint256[] values, bytes[] calldatas, bytes32 descriptionHash) returns (uint256)',
  'function execute(address[] targets, uint256[] values, bytes[] calldatas, bytes32 descriptionHash) returns (uint256)',
  'function cancel(address[] targets, uint256[] values, bytes[] calldatas, bytes32 descriptionHash) returns (uint256)',
  
  // Voting functions
  'function castVote(uint256 proposalId, uint8 support) returns (uint256)',
  'function castVoteWithReason(uint256 proposalId, uint8 support, string reason) returns (uint256)',
  'function castVoteWithReasonAndParams(uint256 proposalId, uint8 support, string reason, bytes params) returns (uint256)',
  
  // View functions
  'function state(uint256 proposalId) view returns (uint8)',
  'function proposalSnapshot(uint256 proposalId) view returns (uint256)',
  'function proposalDeadline(uint256 proposalId) view returns (uint256)',
  'function proposalVotes(uint256 proposalId) view returns (uint256 againstVotes, uint256 forVotes, uint256 abstainVotes)',
  'function hasVoted(uint256 proposalId, address account) view returns (bool)',
  'function getVotes(address account, uint256 blockNumber) view returns (uint256)',
  'function quorum(uint256 blockNumber) view returns (uint256)',
  'function votingDelay() view returns (uint256)',
  'function votingPeriod() view returns (uint256)',
  'function proposalThreshold() view returns (uint256)',
  'function name() view returns (string)',
  
  // Events
  'event ProposalCreated(uint256 proposalId, address proposer, address[] targets, uint256[] values, string[] signatures, bytes[] calldatas, uint256 startBlock, uint256 endBlock, string description)',
  'event ProposalCanceled(uint256 proposalId)',
  'event ProposalExecuted(uint256 proposalId)',
  'event VoteCast(address indexed voter, uint256 proposalId, uint8 support, uint256 weight, string reason)',
];

// Proposal States
const ProposalState = {
  0: 'Pending',
  1: 'Active',
  2: 'Canceled',
  3: 'Defeated',
  4: 'Succeeded',
  5: 'Queued',
  6: 'Expired',
  7: 'Executed',
};

// Vote Types
const VoteType = {
  Against: 0,
  For: 1,
  Abstain: 2,
};

class GovernanceService {
  constructor() {
    this.provider = null;
    this.wallet = null;
    this.governor = null;
    this.initialized = false;
  }

  /**
   * Initialize the governance service
   */
  async initialize() {
    if (this.initialized) {
      return;
    }

    try {
      // Setup provider
      this.provider = new ethers.JsonRpcProvider(config.blockchain.rpcUrl);
      
      // Setup wallet
      this.wallet = new ethers.Wallet(config.blockchain.privateKey, this.provider);
      
      // Setup Governor contract
      const governorAddress = config.blockchain.governorAddress;
      if (governorAddress && governorAddress !== '0x0000000000000000000000000000000000000000') {
        this.governor = new ethers.Contract(governorAddress, GOVERNOR_ABI, this.wallet);
        logger.info('Governor contract initialized', { address: governorAddress });
      } else {
        logger.warn('Governor address not configured - governance features disabled');
      }

      this.initialized = true;
      logger.info('Governance service initialized successfully');
    } catch (error) {
      logger.error('Failed to initialize governance service', { error: error.message });
      throw error;
    }
  }

  /**
   * Create a new proposal
   * @param {Object} proposalData - Proposal details
   */
  async createProposal(proposalData) {
    await this.initialize();

    if (!this.governor) {
      throw new Error('Governor contract not configured');
    }

    const {
      title,
      description,
      category,
      budgetAmount,
      targets = [],
      values = [],
      calldatas = [],
      proposerId,
      panchayatId,
    } = proposalData;

    try {
      // Build full description with metadata
      const fullDescription = JSON.stringify({
        title,
        description,
        category,
        budgetAmount,
        proposerId,
        panchayatId,
        timestamp: Date.now(),
        version: '1.0',
      });

      // If no specific on-chain actions, create a signaling proposal
      const finalTargets = targets.length > 0 ? targets : [this.governor.target];
      const finalValues = values.length > 0 ? values : [0];
      const finalCalldatas = calldatas.length > 0 ? calldatas : ['0x'];

      logger.info('Creating proposal', { title, category });

      const tx = await this.governor.propose(
        finalTargets,
        finalValues,
        finalCalldatas,
        fullDescription
      );

      const receipt = await tx.wait();
      
      // Extract proposal ID from event
      const proposalCreatedEvent = receipt.logs.find(
        log => log.fragment?.name === 'ProposalCreated'
      );

      const proposalId = proposalCreatedEvent?.args?.proposalId?.toString() || 
                         ethers.keccak256(ethers.toUtf8Bytes(fullDescription));

      logger.info('Proposal created successfully', { 
        proposalId, 
        txHash: receipt.hash 
      });

      return {
        success: true,
        proposalId,
        txHash: receipt.hash,
        title,
        description: fullDescription,
        category,
        status: 'Pending',
      };
    } catch (error) {
      logger.error('Failed to create proposal', { error: error.message });
      throw error;
    }
  }

  /**
   * Cast a vote on a proposal
   * @param {string} proposalId - Proposal ID
   * @param {Object} voteData - Vote details
   */
  async castVote(proposalId, voteData) {
    await this.initialize();

    if (!this.governor) {
      throw new Error('Governor contract not configured');
    }

    const { support, reason = '', voterId } = voteData;

    try {
      // Convert support to number
      let voteType;
      if (typeof support === 'string') {
        voteType = support.toLowerCase() === 'for' ? VoteType.For :
                   support.toLowerCase() === 'against' ? VoteType.Against :
                   VoteType.Abstain;
      } else {
        voteType = support;
      }

      logger.info('Casting vote', { proposalId, support: voteType, voterId });

      let tx;
      if (reason) {
        tx = await this.governor.castVoteWithReason(proposalId, voteType, reason);
      } else {
        tx = await this.governor.castVote(proposalId, voteType);
      }

      const receipt = await tx.wait();

      logger.info('Vote cast successfully', { 
        proposalId, 
        txHash: receipt.hash,
        support: voteType 
      });

      return {
        success: true,
        proposalId,
        txHash: receipt.hash,
        support: voteType,
        reason,
      };
    } catch (error) {
      logger.error('Failed to cast vote', { error: error.message, proposalId });
      throw error;
    }
  }

  /**
   * Get proposal details
   * @param {string} proposalId - Proposal ID
   */
  async getProposal(proposalId) {
    await this.initialize();

    if (!this.governor) {
      throw new Error('Governor contract not configured');
    }

    try {
      const [state, snapshot, deadline, votes] = await Promise.all([
        this.governor.state(proposalId),
        this.governor.proposalSnapshot(proposalId),
        this.governor.proposalDeadline(proposalId),
        this.governor.proposalVotes(proposalId),
      ]);

      return {
        proposalId,
        state: ProposalState[state] || 'Unknown',
        stateCode: Number(state),
        snapshot: snapshot.toString(),
        deadline: deadline.toString(),
        votes: {
          against: votes.againstVotes.toString(),
          for: votes.forVotes.toString(),
          abstain: votes.abstainVotes.toString(),
        },
      };
    } catch (error) {
      logger.error('Failed to get proposal', { error: error.message, proposalId });
      throw error;
    }
  }

  /**
   * Get governance parameters
   */
  async getGovernanceParams() {
    await this.initialize();

    if (!this.governor) {
      return {
        configured: false,
        message: 'Governor contract not configured',
      };
    }

    try {
      const [name, votingDelay, votingPeriod, proposalThreshold] = await Promise.all([
        this.governor.name(),
        this.governor.votingDelay(),
        this.governor.votingPeriod(),
        this.governor.proposalThreshold(),
      ]);

      const currentBlock = await this.provider.getBlockNumber();
      const quorum = await this.governor.quorum(currentBlock - 1);

      return {
        configured: true,
        name,
        votingDelay: votingDelay.toString(),
        votingPeriod: votingPeriod.toString(),
        proposalThreshold: proposalThreshold.toString(),
        quorum: quorum.toString(),
        currentBlock,
      };
    } catch (error) {
      logger.error('Failed to get governance params', { error: error.message });
      throw error;
    }
  }

  /**
   * Execute a passed proposal
   * @param {string} proposalId - Proposal ID
   * @param {Object} proposalData - Original proposal data for hash
   */
  async executeProposal(proposalId, proposalData) {
    await this.initialize();

    if (!this.governor) {
      throw new Error('Governor contract not configured');
    }

    const { targets, values, calldatas, description } = proposalData;

    try {
      const descriptionHash = ethers.keccak256(ethers.toUtf8Bytes(description));

      logger.info('Executing proposal', { proposalId });

      const tx = await this.governor.execute(
        targets,
        values,
        calldatas,
        descriptionHash
      );

      const receipt = await tx.wait();

      logger.info('Proposal executed successfully', { 
        proposalId, 
        txHash: receipt.hash 
      });

      return {
        success: true,
        proposalId,
        txHash: receipt.hash,
        status: 'Executed',
      };
    } catch (error) {
      logger.error('Failed to execute proposal', { error: error.message, proposalId });
      throw error;
    }
  }

  /**
   * Check if an address has voted on a proposal
   * @param {string} proposalId - Proposal ID
   * @param {string} address - Voter address
   */
  async hasVoted(proposalId, address) {
    await this.initialize();

    if (!this.governor) {
      throw new Error('Governor contract not configured');
    }

    try {
      const voted = await this.governor.hasVoted(proposalId, address);
      return { proposalId, address, hasVoted: voted };
    } catch (error) {
      logger.error('Failed to check vote status', { error: error.message });
      throw error;
    }
  }

  /**
   * Get voting power for an address
   * @param {string} address - Voter address
   * @param {number} blockNumber - Block number (optional, uses current - 1)
   */
  async getVotingPower(address, blockNumber = null) {
    await this.initialize();

    if (!this.governor) {
      throw new Error('Governor contract not configured');
    }

    try {
      const block = blockNumber || (await this.provider.getBlockNumber()) - 1;
      const votes = await this.governor.getVotes(address, block);
      return {
        address,
        blockNumber: block,
        votingPower: votes.toString(),
      };
    } catch (error) {
      logger.error('Failed to get voting power', { error: error.message, address });
      throw error;
    }
  }
}

// Export singleton instance
module.exports = new GovernanceService();

File: lib/core/services/web3/governance_service.dart
/// GramPulse Governance Service - DAO Client
///
/// Flutter client for interacting with the governance backend API.
/// Handles proposal creation, voting, and governance queries.

import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:flutter/foundation.dart';
import 'web3_config.dart';

/// Proposal state enum
enum ProposalState {
  pending,
  active,
  canceled,
  defeated,
  succeeded,
  queued,
  expired,
  executed,
  unknown,
}

/// Vote type enum
enum VoteType {
  against,
  forVote,
  abstain,
}

/// Proposal model
class Proposal {
  final String id;
  final String title;
  final String description;
  final String category;
  final double? budgetAmount;
  final String? proposerId;
  final String? panchayatId;
  final ProposalState state;
  final String? snapshot;
  final String? deadline;
  final ProposalVotes? votes;
  final String? txHash;
  final DateTime? createdAt;

  Proposal({
    required this.id,
    required this.title,
    required this.description,
    required this.category,
    this.budgetAmount,
    this.proposerId,
    this.panchayatId,
    this.state = ProposalState.pending,
    this.snapshot,
    this.deadline,
    this.votes,
    this.txHash,
    this.createdAt,
  });

  factory Proposal.fromJson(Map<String, dynamic> json) {
    return Proposal(
      id: json['proposalId']?.toString() ?? json['id']?.toString() ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      category: json['category'] ?? '',
      budgetAmount: json['budgetAmount']?.toDouble(),
      proposerId: json['proposerId'],
      panchayatId: json['panchayatId'],
      state: _parseState(json['state'] ?? json['status']),
      snapshot: json['snapshot'],
      deadline: json['deadline'],
      votes: json['votes'] != null ? ProposalVotes.fromJson(json['votes']) : null,
      txHash: json['txHash'],
      createdAt: json['createdAt'] != null 
          ? DateTime.tryParse(json['createdAt']) 
          : null,
    );
  }

  static ProposalState _parseState(dynamic state) {
    if (state == null) return ProposalState.unknown;
    final stateStr = state.toString().toLowerCase();
    switch (stateStr) {
      case 'pending':
      case '0':
        return ProposalState.pending;
      case 'active':
      case '1':
        return ProposalState.active;
      case 'canceled':
      case '2':
        return ProposalState.canceled;
      case 'defeated':
      case '3':
        return ProposalState.defeated;
      case 'succeeded':
      case '4':
        return ProposalState.succeeded;
      case 'queued':
      case '5':
        return ProposalState.queued;
      case 'expired':
      case '6':
        return ProposalState.expired;
      case 'executed':
      case '7':
        return ProposalState.executed;
      default:
        return ProposalState.unknown;
    }
  }

  bool get isVotingActive => state == ProposalState.active;
  bool get isPassed => state == ProposalState.succeeded || state == ProposalState.executed;
  bool get isFailed => state == ProposalState.defeated || state == ProposalState.canceled;
}

/// Proposal votes model
class ProposalVotes {
  final BigInt against;
  final BigInt forVotes;
  final BigInt abstain;

  ProposalVotes({
    required this.against,
    required this.forVotes,
    required this.abstain,
  });

  factory ProposalVotes.fromJson(Map<String, dynamic> json) {
    return ProposalVotes(
      against: BigInt.tryParse(json['against']?.toString() ?? '0') ?? BigInt.zero,
      forVotes: BigInt.tryParse(json['for']?.toString() ?? '0') ?? BigInt.zero,
      abstain: BigInt.tryParse(json['abstain']?.toString() ?? '0') ?? BigInt.zero,
    );
  }

  BigInt get total => against + forVotes + abstain;
  
  double get forPercentage => total > BigInt.zero 
      ? (forVotes.toDouble() / total.toDouble()) * 100 
      : 0;
  
  double get againstPercentage => total > BigInt.zero 
      ? (against.toDouble() / total.toDouble()) * 100 
      : 0;
  
  double get abstainPercentage => total > BigInt.zero 
      ? (abstain.toDouble() / total.toDouble()) * 100 
      : 0;
}

/// Governance parameters model
class GovernanceParams {
  final bool configured;
  final String? name;
  final String? votingDelay;
  final String? votingPeriod;
  final String? proposalThreshold;
  final String? quorum;
  final int? currentBlock;

  GovernanceParams({
    required this.configured,
    this.name,
    this.votingDelay,
    this.votingPeriod,
    this.proposalThreshold,
    this.quorum,
    this.currentBlock,
  });

  factory GovernanceParams.fromJson(Map<String, dynamic> json) {
    return GovernanceParams(
      configured: json['configured'] ?? false,
      name: json['name'],
      votingDelay: json['votingDelay'],
      votingPeriod: json['votingPeriod'],
      proposalThreshold: json['proposalThreshold'],
      quorum: json['quorum'],
      currentBlock: json['currentBlock'],
    );
  }
}

/// Vote result model
class VoteResult {
  final bool success;
  final String proposalId;
  final String? txHash;
  final VoteType support;
  final String? reason;

  VoteResult({
    required this.success,
    required this.proposalId,
    this.txHash,
    required this.support,
    this.reason,
  });

  factory VoteResult.fromJson(Map<String, dynamic> json) {
    return VoteResult(
      success: json['success'] ?? false,
      proposalId: json['proposalId']?.toString() ?? '',
      txHash: json['txHash'],
      support: VoteType.values[json['support'] ?? 1],
      reason: json['reason'],
    );
  }
}

/// Governance Service for DAO operations
class GovernanceService {
  final Web3Config _config;
  final http.Client _client;
  
  GovernanceService({
    Web3Config? config,
    http.Client? client,
  }) : _config = config ?? Web3Config.fromEnvironment(),
       _client = client ?? http.Client();

  String get _baseUrl => _config.backendUrl;
  
  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    'x-api-key': _config.backendApiKey,
  };

  /// Create a new proposal
  Future<Proposal> createProposal({
    required String title,
    required String description,
    required String category,
    double? budgetAmount,
    String? proposerId,
    String? panchayatId,
  }) async {
    try {
      final response = await _client.post(
        Uri.parse('$_baseUrl/governance/proposal'),
        headers: _headers,
        body: jsonEncode({
          'title': title,
          'description': description,
          'category': category,
          if (budgetAmount != null) 'budgetAmount': budgetAmount,
          if (proposerId != null) 'proposerId': proposerId,
          if (panchayatId != null) 'panchayatId': panchayatId,
        }),
      );

      if (response.statusCode == 201) {
        final data = jsonDecode(response.body);
        return Proposal.fromJson(data['data']);
      } else {
        final error = jsonDecode(response.body);
        throw GovernanceException(error['error'] ?? 'Failed to create proposal');
      }
    } catch (e) {
      if (e is GovernanceException) rethrow;
      throw GovernanceException('Network error: $e');
    }
  }

  /// Cast a vote on a proposal
  Future<VoteResult> castVote({
    required String proposalId,
    required VoteType support,
    String? reason,
    String? voterId,
  }) async {
    try {
      String supportStr;
      switch (support) {
        case VoteType.against:
          supportStr = 'against';
          break;
        case VoteType.forVote:
          supportStr = 'for';
          break;
        case VoteType.abstain:
          supportStr = 'abstain';
          break;
      }

      final response = await _client.post(
        Uri.parse('$_baseUrl/governance/vote'),
        headers: _headers,
        body: jsonEncode({
          'proposalId': proposalId,
          'support': supportStr,
          if (reason != null) 'reason': reason,
          if (voterId != null) 'voterId': voterId,
        }),
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return VoteResult.fromJson(data['data']);
      } else {
        final error = jsonDecode(response.body);
        throw GovernanceException(error['error'] ?? 'Failed to cast vote');
      }
    } catch (e) {
      if (e is GovernanceException) rethrow;
      throw GovernanceException('Network error: $e');
    }
  }

  /// Get proposal details
  Future<Proposal> getProposal(String proposalId) async {
    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/governance/proposal/$proposalId'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return Proposal.fromJson(data['data']);
      } else {
        final error = jsonDecode(response.body);
        throw GovernanceException(error['error'] ?? 'Failed to get proposal');
      }
    } catch (e) {
      if (e is GovernanceException) rethrow;
      throw GovernanceException('Network error: $e');
    }
  }

  /// Get governance parameters
  Future<GovernanceParams> getGovernanceParams() async {
    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/governance/params'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return GovernanceParams.fromJson(data['data']);
      } else {
        final error = jsonDecode(response.body);
        throw GovernanceException(error['error'] ?? 'Failed to get governance params');
      }
    } catch (e) {
      if (e is GovernanceException) rethrow;
      throw GovernanceException('Network error: $e');
    }
  }

  /// Check if an address has voted on a proposal
  Future<bool> hasVoted(String proposalId, String address) async {
    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/governance/voted/$proposalId/$address'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['data']['hasVoted'] ?? false;
      } else {
        return false;
      }
    } catch (e) {
      debugPrint('Error checking vote status: $e');
      return false;
    }
  }

  /// Get voting power for an address
  Future<BigInt> getVotingPower(String address) async {
    try {
      final response = await _client.get(
        Uri.parse('$_baseUrl/governance/voting-power/$address'),
        headers: {'Content-Type': 'application/json'},
      );

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return BigInt.tryParse(data['data']['votingPower']?.toString() ?? '0') ?? BigInt.zero;
      } else {
        return BigInt.zero;
      }
    } catch (e) {
      debugPrint('Error getting voting power: $e');
      return BigInt.zero;
    }
  }

  void dispose() {
    _client.close();
  }
}

/// Governance exception
class GovernanceException implements Exception {
  final String message;
  GovernanceException(this.message);
  
  @override
  String toString() => 'GovernanceException: $message';
}
